File: swift/Sources/Authup/Webhook.swift
import Foundation
import CryptoKit

/// Handles webhook verification using HMAC-SHA256.
public enum WebhookVerifier {

    /// Verifies a webhook payload against the provided signature.
    ///
    /// - Parameters:
    ///   - payload: The raw webhook payload bytes.
    ///   - signature: The hex-encoded HMAC-SHA256 signature.
    ///   - secret: The webhook signing secret.
    /// - Returns: The parsed JSON dictionary.
    /// - Throws: `WebhookError` if verification fails.
    public static func verify(
        payload: Data,
        signature: String,
        secret: String
    ) throws -> [String: Any] {
        guard !payload.isEmpty else {
            throw WebhookError.emptyPayload
        }
        guard !signature.isEmpty else {
            throw WebhookError.emptySignature
        }
        guard !secret.isEmpty else {
            throw WebhookError.emptySecret
        }

        let key = SymmetricKey(data: Data(secret.utf8))
        let mac = HMAC<SHA256>.authenticationCode(for: payload, using: key)
        let expected = mac.map { String(format: "%02x", $0) }.joined()

        guard expected == signature else {
            throw WebhookError.invalidSignature
        }

        guard let json = try JSONSerialization.jsonObject(with: payload) as? [String: Any] else {
            throw WebhookError.invalidPayload
        }

        return json
    }
}

/// Errors that can occur during webhook verification.
public enum WebhookError: Error, Sendable {
    case emptyPayload
    case emptySignature
    case emptySecret
    case invalidSignature
    case invalidPayload
}

File: kotlin/src/main/kotlin/dev/authup/sdk/Webhook.kt
package dev.authup.sdk

import com.google.gson.Gson
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/** Handles webhook verification using HMAC-SHA256. */
object WebhookVerifier {

    private val gson = Gson()

    /**
     * Verifies a webhook payload against the provided HMAC-SHA256 signature.
     *
     * @param payload The raw webhook payload string.
     * @param signature The hex-encoded HMAC-SHA256 signature.
     * @param secret The webhook signing secret.
     * @return The parsed JSON map.
     * @throws IllegalArgumentException If any parameter is empty.
     * @throws SecurityException If the signature is invalid.
     */
    @JvmStatic
    fun verify(payload: String, signature: String, secret: String): Map<String, Any> {
        require(payload.isNotEmpty()) { "Empty webhook payload" }
        require(signature.isNotEmpty()) { "Empty webhook signature" }
        require(secret.isNotEmpty()) { "Empty webhook secret" }

        val mac = Mac.getInstance("HmacSHA256")
        val keySpec = SecretKeySpec(secret.toByteArray(Charsets.UTF_8), "HmacSHA256")
        mac.init(keySpec)
        val hash = mac.doFinal(payload.toByteArray(Charsets.UTF_8))
        val expected = hash.joinToString("") { "%02x".format(it) }

        if (expected != signature) {
            throw SecurityException("Invalid webhook signature")
        }

        @Suppress("UNCHECKED_CAST")
        return gson.fromJson(payload, Map::class.java) as Map<String, Any>
    }
}
